jlm/llvm/backend/dot/DotWriter.cpp

```diff
@@ -0,0 +1,206 @@
+      auto & edge = typeGraph.CreateDirectedEdge(resultTypeNode, node);
+      edge.SetAttribute("Result#", util::strfmt(n));
+    }
+  }
+  else
+  {
+    JLM_UNREACHABLE("Unknown type");
+  }
+
+  return node;
+}
+
+/**
+ * Attaches the given GraphWriter port to the input in the RVSDG it represents.
+ * Also adds an edge to the input port, from the node representing the input's origin.
+ * @param inputPort the GraphWriter port representing the input
+ * @param rvsdgInput the RVSDG input
+ */
+static void
+AttachNodeInput(util::Port & inputPort, const rvsdg::input & rvsdgInput)
+{
+  auto & graph = inputPort.GetGraph();
+  inputPort.SetProgramObject(rvsdgInput);
+
+  // nodes are visited in topological order, so if the origin is an output, it will already exist
+  if (auto originPort =
+          reinterpret_cast<util::Port *>(graph.GetElementFromProgramObject(*rvsdgInput.origin())))
+  {
+    auto & edge = graph.CreateDirectedEdge(*originPort, inputPort);
+    if (rvsdg::is<MemoryStateType>(rvsdgInput.type()))
+      edge.SetAttribute("color", util::Colors::Red);
+    if (rvsdg::is<iostatetype>(rvsdgInput.type()))
+      edge.SetAttribute("color", util::Colors::Green);
+  }
+}
+
+/**
+ * Attaches the given GraphWriter port to the output in RVSDG it represents.
+ * Also adds information to the output about its type, using a reference to the type graph.
+ * @param outputPort the GraphWriter port representing the output
+ * @param rvsdgOutput the RVSDG output
+ * @param typeGraph the type graph, or nullptr if the output's type should not be included
+ */
+static void
+AttachNodeOutput(
+    util::Port & outputPort,
+    const rvsdg::output & rvsdgOutput,
+    util::Graph * typeGraph)
+{
+  outputPort.SetProgramObject(rvsdgOutput);
+  if (typeGraph)
+    outputPort.SetAttributeGraphElement(
+        "type",
+        GetOrCreateTypeGraphNode(rvsdgOutput.type(), *typeGraph));
+}
+
+/**
+ * Fill the given \p graph with nodes corresponding to the nodes of the given \p region.
+ * If \p typeGraph is not nullptr, all rvsdg outputs get a type reference to the type graph.
+ * If the type does not already exist in the type graph, it is created.
+ */
+static void
+CreateGraphNodes(util::Graph & graph, rvsdg::region & region, util::Graph * typeGraph)
+{
+  graph.SetProgramObject(region);
+
+  // Start by creating nodes for all the region arguments, and attaching them to the RVSDG outputs.
+  for (size_t n = 0; n < region.narguments(); n++)
+  {
+    auto & node = graph.CreateArgumentNode();
+    auto & argument = *region.argument(n);
+    AttachNodeOutput(node, argument, typeGraph);
+
+    // If this argument corresponds to one of the structural node's inputs, reference it
+    if (argument.input())
+      node.SetAttributeObject("input", *argument.input());
+  }
+
+  // Create a node for each node in the region
+  rvsdg::topdown_traverser traverser(&region);
+  for (const auto rvsdgNode : traverser)
```

Comment:
Does the order of traversal make a difference here? Could you not just use `(const auto rvsdgNode : region.nodes)`?

Reply:
It does make a difference, as inputs need to add edges to the output representing the origin, which has to be created first. Will make it clearer